Automated install tests

ydirson opened the pull request:
Now based on top of preliminary PRs to separate generally-useful stuff:
* #242
* #250 
* #251 
* #252 
* https://github.com/xcp-ng/xcp-ng-tests/pull/283

Requires at least python 3.8 and pytest 8.2.

gduperrey commented on scripts/install_xcpng.py:
```diff
@@ -237,7 +198,7 @@ def main():
         generate_answerfile(tmp_local_path, installer, args.host, args.target_hostname, args.action, hdd,
                             netinstall_gpg_check)
-        generate_boot_conf(tmp_local_path, installer, args.action)
+        pxe.generate_boot_conf(tmp_local_path, installer)
```
I don't understand this change. This option was used to have 3 actions: install, upgrade or restore.
I don't see how is it managed in the pxe file. It doesn't seems to exist anymore.

ydirson replied:
`generate_boot_conf` only uses the action to set `rt=1` as a workaround in the `restore` case, I want to have a look at that to see if we can do better (e.g. an answerfile solution)

ydirson replied:
this was reverted and split into #252

gduperrey commented on data.py-dist:
```diff
@@ -26,6 +26,15 @@ PXE_CONFIG_SERVER = 'pxe'
 # Default VM images location
 DEF_VM_URL = 'http://pxe/images/'
 
+# Default shared ISO SR
```
Why using another ISO SR when we already have `DEFAULT_NFS_ISO_DEVICE_CONFIG` in the data.py?

ydirson replied:
Ah, I had missed that one

ydirson replied:
In fact those are different, `DEFAULT_NFS_ISO_DEVICE_CONFIG` is used in tests that create a new ISO SR from an existing NFS server, while what I'm doing is using an existing ISO SR.

Using an existing shared ISO SR is in fact not perfect because of the possibility of interference with other users (filename clashes), but turning an existing NFS share into an ISO SR does not help this, and does not seem really suitable when working with hosts that already have an ISO SR configured.

OTOH the test could provide its own ISO SR to workaround the filename clashes - or, more comfortably, make a clever use of `mkstemp` through `ssh`.

ydirson replied:
This was changed recently: we now use the master's ISO-SR mountpoint, no need for an extra config.

ydirson commented:
Current status: install + firstboot work in a single test
Next:
- xva export/import, split firstboot in separate test
- include answerfile in `install.img`
- answerfile generation

ydirson commented:
Current status: 2 separate tests for install and firstboot work without any manual intervention
Next step: write an upgrade tests

ydirson commented:
Current status:
- 3 separate tests for install, firstboot, "reinstall upgrade"
- test dependencies declared (requires `pytest-dependency`, best used with `pytest-order`)
- includes `pytest_collection_modifyitems` hook to automatically run the dependencies first

Next steps:
- [ ] improve `pytest_collection_modifyitems` hook by not rerunning dependencies whose output is in the vm cache
- [ ] improve hook to deal with test parameters
- [ ] start adding (explicit) parameters on tests
- [ ] improve hook further by not ignoring dependents of tests explicitly requested (so in `a->b->c->d` chain with cached VMs `pytest b d` will also run `c`

ydirson commented:
Current status:
- improvements to test parametrization well under way, but bad interations with dependency, to be investigated

gthvn1 commented on conftest.py:
```diff
@@ -331,6 +378,159 @@ def imported_vm(host, vm_ref):
+
+    import git
+    test_repo = git.Repo(".")
+    assert not test_repo.is_dirty(), "test repo must not be dirty to cache images"
```
should not be tested if CACHE_IMPORTED_VM is set to False no?

ydirson replied:
Not applicable any more: CACHE_IMPORTED_VM now has no impact on nested-host caching, which is systematic

gthvn1 commented on data.py-dist:
```diff
@@ -20,18 +24,42 @@ HOSTS = {
+#    'xcpng-8.2.0': "/home/user/iso/xcp-ng-8.2.0.iso",
+#    'xcpng-8.2.1': "/home/user/iso/xcp-ng-8.2.1.iso",
+#    'xcpng-8.2.1-2023': "/home/user/iso/xcp-ng-8.2.1-20231130.iso",
+#    'xcpng-8.3-beta2': "/home/user/iso/xcp-ng-8.3.0-beta2.iso",
```
maybe `xcpng-8.3.0-beta2` ? because we will have at some point an `xcpng-8.3.1-beta2`... It looks like a bit more consistent IMHO.

ydirson commented:
Last push brings:
* `data.py` impact:
  * management network specified by name not uuid
  * install hosts with static IPs from `data.py` instead of mis-using the DHCP range
  * answerfile generation completely reworked for better flexibility
  * (WIP) one single nickname for images instead of 2 layers
* starting up a host with an install CD is now a fixture
* python requirement bumped to 3.8
* new CI job to test test-sequences contents a bit
* added test sequences for upgrade from 8.3beta2 (known broken in rc1)
* robustness fixes

ydirson commented:
rebased

ydirson commented:
Added support for tests on nightly ISOs

ydirson commented:
Latest push: assorted fixes + foundations for testing `pool-join` after upgrade

ydirson commented:
Latest push: more doc, history cleanup and fixes (had obviously-broken intermediate commits)

ydirson commented:
Latest push:
* split out RPU for later into #253 
* redo test-sequences as incremental, with prerequisite inside filename; move most of them to a single last commit
* use `83nightly` not just `nightly` since we'll want some `82nightly` at one point
* rationalised combos: `83nightly` is the sole ISO version used used for upgrading and restoring for now

ydirson commented:
Latest push:
- a few fixes around `IMAGE_DEFAULT_EQUIVS`
- rebase

ydirson commented:
Latest push:
- properly destroy cache:// hosts when the `hosts` fixture setup errors out
- upload ISO to ISOSR through the nest's pool master, no need to set NFS server and path in `data.py` any more
- UUID regen before firstboot now handled by `test_tune_firstboot` together with hostname and IP, rather than through an additional oneshot service
- rescan only the right ISOSR, and only once after uploading an ISO
- completely get rid of `installer_iso` mark, ISO is directly infered from `iso_version` param
- fixes for runtests scripts

ydirson commented:
Latest push:
- rebased
- renaming of netinstall test param value to `83rcnet`
- some generalisations to cope with some xs65 aspects
- small refactoring with `install_disk` fixture
- a few more cleanups and bugfixes
- split fsdiff work into a separate branch
- split runtests scripts into a separate branch

ydirson commented:
Latest push:
- fixed last issues causing commits to be marked WIP
- added some consistency to `install/test.py`
- got rid of `param_mapping` which was essentially used to specify 1:1 mappings and felt like just overhead

ydirson commented:
Latest push:
- cleaned up unused imports
- renamed `source_type` to `package_source` for netinstall support

stormi commented:
@benjamreis This PR is now ready, let's review it in the next days.

ydirson commented:
Last push fixes the check of installed OS name after upgrade

ydirson commented:
a few fixups:
- better log
- fix firstboot version check for successive-upgrade tests (otherwise not included in that branch)
- sort checks for firstboot stamps to match their creation ordering

ydirson commented:
And last fixup: the long-talked-about "back to DHCP"

gthvn1 reviewed (approved):
LGTM, 
hats off for the PR (french translation so I hope it makes sense :) )

ydirson commented:
Rebased on latest master, while apply fixups to relevant commits.  And fixed the doubled prefix in the VM label on initial tests in a chain.

ydirson commented:
Fixed `pycodestyle` complaints

ydirson commented:
Latest pushes:
* enable `82nightly` and `83nightly` in default config, in a form where they require an envvar to point the specific snapshot image to be used
* drop obsolete 8.3 prereleases and provide 8.3.0 instead (as a base for nightly upgrade tests)
* provide a first README

glehmann commented on tests/install/README.md:
```diff
@@ -0,0 +1,176 @@
+inside a pre-existing XCP-ng pool.  In standard terminology of
+hypervisor nesting, we have:
+
+L0: the pre-existing XCP-ng pool
+L1: the XCP-ng host(s) exercised by the tests
+L2: VMs launched by the hosts under test
```
```suggestion
* L0: the pre-existing XCP-ng pool
* L1: the XCP-ng host(s) exercised by the tests
* L2: VMs launched by the hosts under test
```

ydirson replied:
fix pushed

ydirson commented:
Last push: uncomment most of `ISO_IMAGES` from `data.py-dist`, it is usable as-is since we're not limited to local files any more.

ydirson commented:
Changes since 920aa72f2fc5: rebase onto new master with mypy checking activated require a few adjustments

glehmann commented on conftest.py:
```diff
@@ -137,22 +148,71 @@ def pytest_runtest_makereport(item, call):
 def hosts(pytestconfig):
+    nested_list = []
+
+    def setup_host(hostname_or_ip, *, config=None):
+        host_vm = None
+        if hostname_or_ip.startswith("cache://"):
+            nest_hostname = config.getoption("nest")
```
There is no check to ensure that `config` is not `None`.
In fact `setup_host` is defined and used only inside `hosts()` and could as well use `pytestconfig` from `hosts()` instead of passing the config as parameter.

ydirson replied:
Explicit check added

ydirson commented:
Latest push:
* added a few commits previously erroneously only in `install-runtest`:
  * support the new nested-virt setting for a 8.3 L0
  * separate `ARP_SERVER` from `PXE_CONFIG_SERVER`
  * support for Alpine as a `ARP_SERVER`
  * some more doc

ydirson commented:
Rebased on current master, fixed a couple of `py{doc,code}style` reports

ydirson commented:
Addressed last review comment

ydirson commented:
Added missing `boot_rst ` configurations and fixed config ordering inconsistencies

ydirson commented:
Removed a stray invalid `boot_rst` config

ydirson commented:
Last push fixes:
- `test_pool_join` not ejecting the 2nd host before teardown
- `test_upgrade[net]` trying to use the local repo
- import consistency in `data.py-dist`

stormi reviewed (approved):
:tada:

ydirson commented:
Rebased on latest master and squashed the fixups

ydirson commented:
Summary of where we stand:
* manual subset of install tests passed:
  - (10 jobs) `-k "uefi-821.1 and iso-nosr and not host2 and not 82nightly`
  - (7 jobs) `-k "bios-830- and iso-ext and not host2"`
  - (7 jobs) `-k "uefi-830net and net-lvm and not host2"`
  - (11 jobs) `-k "uefi-83nightly and iso-nosr"`
  - (7 jobs) `-k "bios-xs8 and iso-nosr and not host2"` (using an old xs8 install iSO :wink:)
* Jenkins 8.2 passes
* regression fixed:
  * jobs-check passes
  * VM prefixes are as expected

We expect the Jenkins 8.3 run to be OK as well, will check at the same time as the next batch.
